File: dotnet/src/Agents/Runtime/Core/TypeSubscription.cs
﻿// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Diagnostics.CodeAnalysis;

namespace Microsoft.SemanticKernel.Agents.Runtime.Core;

/// <summary>
/// This subscription matches on topics based on the exact type and maps to agents using the source of the topic as the agent key.
/// This subscription causes each source to have its own agent instance.
/// </summary>
/// <remarks>
/// Example:
/// <code>
/// var subscription = new TypeSubscription("t1", "a1");
/// </code>
/// In this case:
/// - A <see cref="TopicId"/> with type `"t1"` and source `"s1"` will be handled by an agent of type `"a1"` with key `"s1"`.
/// - A <see cref="TopicId"/> with type `"t1"` and source `"s2"` will be handled by an agent of type `"a1"` with key `"s2"`.
/// </remarks>
public class TypeSubscription : ISubscriptionDefinition
{
    /// <summary>
    /// Initializes a new instance of the <see cref="TypeSubscription"/> class.
    /// </summary>
    /// <param name="topicType">The exact topic type to match against.</param>
    /// <param name="agentType">Agent type to handle this subscription.</param>
    /// <param name="id">Unique identifier for the subscription. If not provided, a new UUID will be generated.</param>
    public TypeSubscription(string topicType, AgentType agentType, string? id = null)
    {
        this.TopicType = topicType;
        this.AgentType = agentType;
        this.Id = id ?? Guid.NewGuid().ToString();
    }

    /// <summary>
    /// Gets the unique identifier of the subscription.
    /// </summary>
    public string Id { get; }

    /// <summary>
    /// Gets the exact topic type used for matching.
    /// </summary>
    public string TopicType { get; }

    /// <summary>
    /// Gets the agent type that handles this subscription.
    /// </summary>
    public AgentType AgentType { get; }

    /// <summary>
    /// Checks if a given <see cref="TopicId"/> matches the subscription based on an exact type match.
    /// </summary>
    /// <param name="topic">The topic to check.</param>
    /// <returns><c>true</c> if the topic's type matches exactly, <c>false</c> otherwise.</returns>
    public bool Matches(TopicId topic)
    {
        return topic.Type == this.TopicType;
    }

    /// <summary>
    /// Maps a <see cref="TopicId"/> to an <see cref="AgentId"/>. Should only be called if <see cref="Matches"/> returns true.
    /// </summary>
    /// <param name="topic">The topic to map.</param>
    /// <returns>An <see cref="AgentId"/> representing the agent that should handle the topic.</returns>
    /// <exception cref="InvalidOperationException">Thrown if the topic does not match the subscription.</exception>
    public AgentId MapToAgent(TopicId topic)
    {
        if (!this.Matches(topic))
        {
            throw new InvalidOperationException("TopicId does not match the subscription.");
        }

        return new AgentId(this.AgentType, topic.Source);
    }

    /// <summary>
    /// Determines whether the specified object is equal to the current subscription.
    /// </summary>
    /// <param name="obj">The object to compare with the current instance.</param>
    /// <returns><c>true</c> if the specified object is equal to this instance; otherwise, <c>false</c>.</returns>
    public override bool Equals([NotNullWhen(true)] object? obj)
    {
        return
            obj is TypeSubscription other &&
                (this.Id == other.Id ||
                    (this.AgentType == other.AgentType &&
                        this.TopicType == other.TopicType));
    }

    /// <summary>
    /// Determines whether the specified subscription is equal to the current subscription.
    /// </summary>
    /// <param name="other">The subscription to compare.</param>
    /// <returns><c>true</c> if the subscriptions are equal; otherwise, <c>false</c>.</returns>
    public bool Equals(ISubscriptionDefinition? other) => this.Id == other?.Id;

    /// <summary>
    /// Returns a hash code for this instance.
    /// </summary>
    /// <returns>A hash code for this instance, suitable for use in hashing algorithms and data structures.</returns>
    public override int GetHashCode()
    {
        return HashCode.Combine(this.Id, this.AgentType, this.TopicType);
    }
}

File: nodejs/semantic-kernel/agents/runtime/in-process/type-subscription.ts
import { randomUUID } from 'crypto'
import { experimental } from '../../../utils/feature-stage-decorator'
import { AgentId, CoreAgentId } from '../core/agent-id'
import { AgentType } from '../core/agent-type'
import { CantHandleException } from '../core/routed-agent'
import { Subscription } from '../core/subscription'
import { TopicId } from '../core/topic'

/**
 * This subscription matches on topics based on the type and maps to agents.
 *
 * It uses the source of the topic as the agent key. This subscription causes each source to have
 * its own agent instance.
 *
 * Note: This class is marked as 'experimental' and may change in the future.
 */
@experimental
export class TypeSubscription implements Subscription {
  private readonly _id: string
  private readonly _topicType: string
  private readonly _agentType: string

  /**
   * Initialize the TypeSubscription.
   *
   * @param topicType - Topic type to match against
   * @param agentType - Agent type to handle this subscription
   * @param id - Id of the subscription. If not provided, a new id will be generated.
   */
  constructor(topicType: string, agentType: string | AgentType, id?: string) {
    this._topicType = topicType

    if (typeof agentType === 'object' && 'type' in agentType) {
      this._agentType = agentType.type
    } else {
      this._agentType = agentType
    }

    this._id = id ?? randomUUID()
  }

  /**
   * Get the id of the subscription.
   */
  get id(): string {
    return this._id
  }

  /**
   * Get the topic type of the subscription.
   */
  get topicType(): string {
    return this._topicType
  }

  /**
   * Get the agent type of the subscription.
   */
  get agentType(): string {
    return this._agentType
  }

  /**
   * Check if the topic_id matches the subscription.
   */
  isMatch(topicId: TopicId): boolean {
    return topicId.type === this._topicType
  }

  /**
   * Map the topic_id to an agent_id.
   *
   * @throws {CantHandleException} If the TopicId does not match the subscription
   */
  mapToAgent(topicId: TopicId): AgentId {
    if (!this.isMatch(topicId)) {
      throw new CantHandleException('TopicId does not match the subscription')
    }

    return new CoreAgentId(this._agentType, topicId.source)
  }

  /**
   * Check if two subscriptions are equal.
   */
  equals(other: Subscription): boolean {
    if (!(other instanceof TypeSubscription)) {
      return false
    }

    return this.id === other.id || (this.agentType === other.agentType && this.topicType === other.topicType)
  }
}
